feat: add option for topics_to_include and exclude_private

RostiMelk opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**
- Added optional feature for including repositories by topic.
  This is done through a comma separated list in env: `topics_to_include`. 
  All repos matching at least one of the topics will get run.
  To keep thing simple, the rest of the repos are added to the ignoredRepositories array.

- Added optional feature for excluding private repositories.
  This is done through a boolean env: `exclude_private`. 
  Repos that are private gets added to the ignoredRepositories array.

**Results**
![55515880d7d68c5ca5889cbeb93d956c](https://user-images.githubusercontent.com/25268506/119546524-4b3c8900-bd94-11eb-8e61-94f5bcf16616.png)
Here is an example with the new options enabled. I currently have a topic enabled for the repo Hyttenett. The rest of the repos are added to the ignoredRepositories array. 
_Some repositories are blurred for confidentiality._

**Related issue(s)**
#8

RostiMelk commented:
Submitted new commit for cleaning up the new filters. 
I ended up placing them in a new file `lib/filters.js`.

derberg commented on lib/filters.js:
```diff
@@ -0,0 +1,33 @@
+ * Getting list of topics that should be included if topics_to_include is set.
+ * Further on we will get a list of repositories that do not belong to any of the specified topics.
+ * 
+ * @param  {Array} topicsToInclude Array of topics to inlcude.
```
it is string on input

RostiMelk commented:
Thanks for the great feedback!

> This is an amazing additional and high-quality PR man! I left only a few comments. Can you also please extend one of examples from readme with these new options https://github.com/derberg/global-workflows-support#advanced-workflow 🙏🏼
> 
> side question, `private` repos, is there a property for it too? when the list of repos is fetched? can I easily determine if repo is public or private?
> 
Yes. You are able to check whether a repo is private or not in the API call. I did a quick test here:
```js
{
      id: 371025676,
      node_id: 'MDEwOlJlcG9zaXRvcnkzNzEwMjU2NzY=',
      name: 'wpe-template-helper',
      full_name: 'designcontainer/wpe-template-helper',
      private: false,
      ...
}
```
> > With new filters being added, I think it is a good idea to move all the filters to its own class. That way the run function doesn't get as cluttered. What do you think
> 
> Tbh, privately I'm not a fan of classes 😄 I use them only if they are really needed, otherwise composition with functions everywhere 😄 I agree though that the core of this action is not easy to digest. I chose to focus on good comments and super detailed logs. I never expected I will have to maintain this action for so long. I hoped to archive it over the native functionality provided by GitHub.....but looks like the adoption grows and even external code contributions are getting in 🎉 And if there will be a need to switch to classes because of it, let it be 😄

Yeah, A class might not be it. A new file was sufficient imo. 👍

RostiMelk commented on action.yml:
```diff
@@ -37,6 +37,16 @@ inputs:
+      Repos that do not contain one of the specified topics will get appended to the repos_to_ignore list. 
+      In the format topic1,topic2.
+    required: false
+  exclude_archived:
```
Do you mean setting topics_to_include by default? I don't recommend that as it will affect older existing setups, and exclude all their repositories.

derberg replied:
tbh I don't remember what was this comment about 😅

RostiMelk commented on lib/index.js:
```diff
@@ -26,18 +27,15 @@ async function run() {
-    //by default repo where workflow runs should always be ignored
-    ignoredRepositories.push(repo);
+    const myOctokit = new octokit(getOctokitOptions(gitHubKey, {
+      previews: ['mercy-preview'],
```
Yes. Topics are currently only available using this preview. I'll add a comment.

RostiMelk commented:
I ended up moving the filters to utils.js, since they caused a loop of some sort when export functions.

Fixes that you mentioned:
- Removed default exclude_archived in readme.md.
- Fixed incorrect arg type in comment.
- Moved ignore filters to utils.js.
- Added comment to why `mercy-review` is used.
- Added example of new filters to advanced workflow in readme.md

I have however not added any default state to topics. As I mentioned in a comment earlier; This will affect older existing setups, and exclude all their repositories.

derberg reviewed (commented):
Code wise perfect! 2 minor comments.

While I was testing manually, I realized that we should not really enable `exclude_archived`, right? If the user will do `exclude_archived: false` or just not provide it at all, the action will fail because technically you cannot push to archived repo anyway, therefore no way to open a PR. I think `exclude_archived` should not be an option but by default archived repos should be ignored.

The code is already there, so maybe instead of removing/refactoring, just change to `exclude_private`? and for archived just have constant exclude in addition?

RostiMelk commented:
Since the GitHub Workflow inputs are treated as strings and not booleans, I had some issues where it would default to true even if false was set. 

This was caused by using `!!`.

Here is a quick explanation of why it shouldn't be used:

> Any string which isn't the empty string will evaluate to true by using them. Although they're the cleanest methods I can think of concerning to boolean conversion, I think they're not what you're looking for.

_Source: https://stackoverflow.com/questions/263965/how-can-i-convert-a-string-to-boolean-in-javascript#answer-264037_

RostiMelk commented on action.yml:
```diff
@@ -37,6 +37,16 @@ inputs:
+    required: false
+  exclude_archived:
+    description: >
+      Boolean value on whether to exclude arcived repositories from this action.
```
Removed now anyway 😅

RostiMelk commented on README.md:
```diff
@@ -46,6 +46,8 @@ committer_username | The username (not display name) of the committer will be us
 commit_message | It is used as a commit message when pushing changes with global workflows. It is also used as a title of the pull request that is created by this action. | false | `Update global workflows`
 repos_to_ignore | Comma-separated list of repositories that should not get updates from this action. Action already ignores the repo in which the action is triggered so you do not need to add it explicitly. In the format `repo1,repo2`. | false | -
+topics_to_include | Comma-separated list of topics that should get updates from this action. Repos that do not contain one of the specified topics will get appended to the repos_to_ignore list. In the format `topic1,topic2`. | false | -
+exclude_archived | Boolean value on whether to exclude arcived repositories from this action. | false | -
```
I agree on this! 
I have written default as false for `exclude_private`.

derberg commented:
looks perfect, just update PR title and description and we are ready to merge and I'll release immediately
